File: app/src/main/java/wxm/KeepAccount/ui/login/FrgLogin.java
package wxm.KeepAccount.ui.login;


import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.TargetApi;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import butterknife.BindString;
import butterknife.BindView;
import wxm.KeepAccount.R;
import wxm.KeepAccount.define.GlobalDef;
import wxm.KeepAccount.ui.usr.ACAddUsr;
import wxm.KeepAccount.ui.welcome.ACWelcome;
import wxm.KeepAccount.utility.ContextUtil;
import wxm.KeepAccount.utility.ToolUtil;
import wxm.androidutil.FrgUtility.FrgSupportBaseAdv;

/**
 * for login
 * Created by WangXM on 2016/11/29.
 */
public class FrgLogin extends FrgSupportBaseAdv {
    // for ui
    @BindView(R.id.login_progress)
    ProgressBar mPBLoginProgress;

    @BindView(R.id.email)
    AutoCompleteTextView mETEmail;

    @BindView(R.id.password)
    EditText mETPassword;

    @BindView(R.id.email_sign_in_button)
    Button mBTEmailSignIn;

    @BindView(R.id.email_register_button)
    Button mBTEmailRegister;

    @BindView(R.id.bt_def_usr_login)
    Button mBTDefUsrLogin;

    @BindView(R.id.ll_login)
    LinearLayout mLLLogin;

    // for res
    @BindString(R.string.error_incorrect_password)
    String mHSErrorPassword;

    @Override
    protected void initUI(Bundle bundle)    {
        mBTEmailSignIn.setOnClickListener(view1 -> attemptLogin());

        mBTEmailRegister.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), ACAddUsr.class);
            startActivityForResult(intent, 1);
        });

        mBTDefUsrLogin.setOnClickListener(v -> doLogin(GlobalDef.DEF_USR_NAME, GlobalDef.DEF_USR_PWD));
    }

    @Override
    protected int getLayoutID() {
        return R.layout.vw_login;
    }

    @Override
    protected boolean isUseEventBus() {
        return false;
    }

    @Override
    protected void loadUI(Bundle bundle) {
    }

    /// PRIVATE BEGIN

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid mETEmail, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private void attemptLogin() {
        // Reset errors.
        mETEmail.setError(null);
        mETPassword.setError(null);

        // Store values at the time of the login attempt.
        String email = mETEmail.getText().toString();
        String password = mETPassword.getText().toString();

        boolean cancel = false;
        View focusView = null;

        // Check for a valid password, if the user entered one.
        if (!TextUtils.isEmpty(password) && !isPasswordValid(password)) {
            mETPassword.setError(getString(R.string.error_invalid_password));
            focusView = mETPassword;
            cancel = true;
        }

        // Check for a valid mETEmail address.
        if (TextUtils.isEmpty(email)) {
            mETEmail.setError(getString(R.string.error_field_required));
            focusView = mETEmail;
            cancel = true;
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView.requestFocus();
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            doLogin(email, password);
        }
    }

    private boolean isPasswordValid(String password) {
        //TODO: Replace this with your own logic
        return password.length() >= 4;
    }


    /**
     * Shows the progress UI and hides the login form.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR2)
    private void showProgress(final boolean show) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        int shortAnimTime = getResources()
                .getInteger(android.R.integer.config_shortAnimTime);

        mLLLogin.setVisibility(show ? View.GONE : View.VISIBLE);
        mLLLogin.animate().setDuration(shortAnimTime).alpha(
                show ? 0 : 1).setListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mLLLogin.setVisibility(show ? View.GONE : View.VISIBLE);
            }
        });

        mPBLoginProgress.setVisibility(show ? View.VISIBLE : View.GONE);
        mPBLoginProgress.animate().setDuration(shortAnimTime)
                .alpha(show ? 1 : 0).setListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mPBLoginProgress.setVisibility(show ? View.VISIBLE : View.GONE);
            }
        });
    }

    /**
     * do login in background
     * @param usr       usr name
     * @param pwd       usr password
     */
    void doLogin(String usr, String pwd)  {
        showProgress(true);

        final boolean[] bret = {false};
        ToolUtil.INSTANCE.runInBackground(this.getActivity(),
                () -> bret[0] = ContextUtil.Companion.getUsrUtility().loginByUsr(usr, pwd),
                () -> {
                    new Handler().postDelayed(() -> showProgress(false), 50);
                    if (bret[0]) {
                        Intent intent = new Intent(getActivity(), ACWelcome.class);
                        startActivityForResult(intent, 1);
                    } else {
                        mETPassword.setError(mHSErrorPassword);
                        mETPassword.requestFocus();
                    }
                });
    }
    /// PRIVATE END
}

File: app/src/main/java/wxm/KeepAccount/ui/login/FrgLogin.kt
package wxm.KeepAccount.ui.login


import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.TargetApi
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.*
import kotterknife.bindView
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import wxm.KeepAccount.R
import wxm.KeepAccount.db.LoginHistoryUtility
import wxm.KeepAccount.define.GlobalDef
import wxm.KeepAccount.event.DoLogin
import wxm.KeepAccount.ui.usr.ACAddUsr
import wxm.KeepAccount.ui.utility.NoteDataHelper
import wxm.KeepAccount.ui.welcome.ACWelcome
import wxm.KeepAccount.utility.AppUtil
import wxm.KeepAccount.utility.ToolUtil
import wxm.androidutil.improve.let1
import wxm.androidutil.log.TagLog
import wxm.androidutil.time.toTimestamp
import wxm.androidutil.ui.frg.FrgSupportBaseAdv
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * for login
 * Created by WangXM on 2016/11/29.
 */
class FrgLogin : FrgSupportBaseAdv() {
    private val mPBLoginProgress: ProgressBar  by bindView(R.id.login_progress)
    private val mETEmail: AutoCompleteTextView by bindView(R.id.email)
    private val mETPassword: EditText by bindView(R.id.password)
    private val mBTEmailSignIn: Button by bindView(R.id.email_sign_in_button)
    private val mBTEmailRegister: Button by bindView(R.id.email_register_button)
    private val mBTDefUsrLogin: Button by bindView(R.id.bt_def_usr_login)
    private val mLLLogin: LinearLayout by bindView(R.id.ll_login)

    override fun getLayoutID(): Int = R.layout.frg_login
    override fun isUseEventBus(): Boolean = true

    /**
     * handler for DB data change
     * @param event     for event
     */
    @Suppress("UNUSED_PARAMETER", "unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onDoLoginEvent(event: DoLogin) {
        TagLog.i("sender is ${event.sendName}")
        reInitUI()
    }

    override fun initUI(bundle: Bundle?) {
        mBTEmailSignIn.setOnClickListener { _ -> attemptLogin() }

        mBTEmailRegister.setOnClickListener { _ ->
            startActivityForResult(Intent(activity, ACAddUsr::class.java), 1)
        }

        mBTDefUsrLogin.setOnClickListener { _ -> doLogin(GlobalDef.DEF_USR_NAME, GlobalDef.DEF_USR_PWD) }

        doHistoryLogin()
    }

    /// PRIVATE BEGIN

    private fun doHistoryLogin() {
        val cl = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -30) }
        LoginHistoryUtility.getLastLoginAfter(cl.toTimestamp())?.let1 {
            ToolUtil.callInBackground(
                    { AppUtil.usrUtility.loginByUsr(it, false) }, false,
                    TimeUnit.SECONDS, 3).let1 {
                if (it) {
                    NoteDataHelper.reloadData()
                    startActivityForResult(Intent(activity, ACWelcome::class.java), 1)
                }
            }
        }
    }

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid mETEmail, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private fun attemptLogin() {
        // Reset errors.
        mETEmail.error = null
        mETPassword.error = null

        // Store values at the time of the login attempt.
        val email = mETEmail.text.toString()
        val password = mETPassword.text.toString()

        var cancel = false
        var focusView: View? = null

        // Check for a valid password, if the user entered one.
        if (!TextUtils.isEmpty(password) && !isPasswordValid(password)) {
            mETPassword.error = getString(R.string.error_invalid_password)
            focusView = mETPassword
            cancel = true
        }

        // Check for a valid mETEmail address.
        if (TextUtils.isEmpty(email)) {
            mETEmail.error = getString(R.string.error_field_required)
            focusView = mETEmail
            cancel = true
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView!!.requestFocus()
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            doLogin(email, password)
        }
    }

    private fun isPasswordValid(password: String): Boolean {
        //TODO: Replace this with your own logic
        return password.length >= 4
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR2)
    private fun showProgress(show: Boolean) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        val shortAnimTime = resources
                .getInteger(android.R.integer.config_shortAnimTime)

        mLLLogin.visibility = if (show) View.GONE else View.VISIBLE
        mLLLogin.animate().setDuration(shortAnimTime.toLong()).alpha(
                (if (show) 0 else 1).toFloat()).setListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                mLLLogin.visibility = if (show) View.GONE else View.VISIBLE
            }
        })

        mPBLoginProgress.visibility = if (show) View.VISIBLE else View.GONE
        mPBLoginProgress.animate().setDuration(shortAnimTime.toLong())
                .alpha((if (show) 1 else 0).toFloat()).setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        mPBLoginProgress.visibility = if (show) View.VISIBLE else View.GONE
                    }
                })
    }

    /**
     * do login in background
     * @param usr       usr name
     * @param pwd       usr password
     */
    private fun doLogin(usr: String, pwd: String) {
        showProgress(true)
        val bRet = ToolUtil.callInBackground(
                { AppUtil.usrUtility.loginByUsr(usr, pwd) }, false,
                TimeUnit.SECONDS, 3)
        showProgress(false)

        if (bRet) {
            NoteDataHelper.reloadData()
            startActivityForResult(Intent(activity, ACWelcome::class.java), 1)
        } else {
            mETPassword.error = getString(R.string.error_incorrect_password)
            mETPassword.requestFocus()
        }
    }
    /// PRIVATE END
}
